### ISSUE TITLE
MPI_COMM_TYPE_SHARED returns NUMA domains, resulting in wrong hvd.local_rank

### ISSUE THREAD

**Issue by surak** [horovod/horovod#3680]
**Environment:**
1. Framework: Tensorflow/PyTorch
2. Framework version: all
3. Horovod version: all
4. MPI version: all
5. CUDA version: all
6. NCCL version: all
7. Python version: 3.*
8. Spark / PySpark version: -
9. Ray version: -
10. OS and version: all
11. GCC version: all
12. CMake version: all

**Checklist:**
1. Did you search issues to find if somebody asked this question before? yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)? It's not
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)? It' not
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? It's not

**Bug report:**
Please describe erroneous behavior you're observing and steps to reproduce it.

The behavior of hvd.local_rank() is depending on a bad assumption.
In mpi_context.cc, the local rank is discovered by creating a split communicator and then getting the rank in this communicator (here: https://github.com/horovod/horovod/blob/a658d0f5315ae175338773b3b93c85e582871ba1/horovod/common/mpi/mpi_context.cc#L102 )

While this works in simpler machines, this will fail in multi-NUMA-domain nodes. The mpi standard states:
> `MPI_COMM_TYPE_SHARED --- this type splits the communicator into subcommunicators, each of which can create a shared memory region.`

So this is not a local rank of one node: this is a single numa domain. 

In such an mpi run (2 nodes, 4 gpus each), where each gpu is connected to a separate numa domain, I get this (the openmpi local rank is correct):

```python
print("DEBUG Horovod Local rank:%d | OpenMPI local rank: %s |  COMM_WORLD rank: %s | Horovod rank: %s" % (hvd.local_rank(),os.environ["OMPI_COMM_WORLD_LOCAL_RANK"], MPI.COMM_WORLD.rank, hvd.rank() ))

DEBUG Horovod Local rank:0 | OpenMPI local rank: 2 |  COMM_WORLD rank: 6 | Horovod rank: 6
DEBUG Horovod Local rank:0 | OpenMPI local rank: 0 |  COMM_WORLD rank: 0 | Horovod rank: 0
DEBUG Horovod Local rank:0 | OpenMPI local rank: 1 |  COMM_WORLD rank: 1 | Horovod rank: 1
DEBUG Horovod Local rank:0 | OpenMPI local rank: 3 |  COMM_WORLD rank: 7 | Horovod rank: 7
DEBUG Horovod Local rank:0 | OpenMPI local rank: 3 |  COMM_WORLD rank: 3 | Horovod rank: 3
DEBUG Horovod Local rank:0 | OpenMPI local rank: 0 |  COMM_WORLD rank: 4 | Horovod rank: 4
DEBUG Horovod Local rank:0 | OpenMPI local rank: 2 |  COMM_WORLD rank: 2 | Horovod rank: 2
DEBUG Horovod Local rank:0 | OpenMPI local rank: 1 |  COMM_WORLD rank: 5 | Horovod rank: 5
```

This breaks, for example, the venerable tf_cnn_benchmark by the Tensorflow guys, as it does something like this:

`config.gpu_options.visible_device_list = str(hvd.local_rank())`
https://github.com/tensorflow/benchmarks/blob/51d647f1962dee3e01a10860dbaf355301ec04da/scripts/tf_cnn_benchmarks/benchmark_cnn.py#L790

This will make all mpi tasks point to gpu 0, which kinda defeats the concept of a local rank.

**Comment by romerojosh** [comment]
I don’t think this is can be a generally true phenomenon. Most multi-GPU systems (e.g. all DGX servers) use CPUs with multiple NUMA domains with GPUs assigned to a single NUMA domain and do not have this issue with getting local rank via splitting with MPI_COMM_TYPE_SHARED. The issue is likely in how your system is configured or how you are launching tasks, which might not allow MPI ranks to use shared memory transports between ranks on the same node, causing this inconsistency.

**Comment by Magnus1cheng** [comment]
I'm facing similiar issues. I'm having 2 nodes, 2 gpus each, On my remote machine: local size = 3, and local ranks are all 2 (which should be 0 and 1.)
How did you solve this issue?

**Comment by Magnus1cheng** [comment]
Just solved this by replace `MPI_Comm_split_type(mpi_comm, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, ` with `MPI_Comm_split_type(mpi_comm, OMPI_COMM_TYPE_NUMA, 0, MPI_INFO_NULL,`. 
I think @romerojosh  is correct, this issue might relates to how system is configured, and maybe how GPUs bind with CPUs.
In my environment, I have 2 CPUs, with 
```
NUMA node0 CPU(s):   0-19,40-59
NUMA node1 CPU(s):   20-39,60-79
````
But I'm wondering if there is a particular reason that Horovod using  MPI_COMM_TYPE_SHARED instead of others. I'm new to MPI so don't really know how to choose the split type.

**Comment by Magnus1cheng** [comment]
Just solved this issue by accident. In my environment, two nodes have exactly same settings (including CPU model), and host name. I changed one of the host name to distinguish these two, and my program starts to running. 
Tried to debug for days, and found the key issue might caused from hwloc module in openmpi, which produces fake info. between process in mpi, so info. like rank, size are all incorrect. Hope this can help you.

**Comment by Magnus1cheng** [comment]
Just solved this issue by accident. In my environment, two nodes have exactly same settings (including CPU model), and host name. I changed one of the host name to distinguish these two, and my program starts to running. 
Tried to debug for days, and found the key issue might caused from hwloc module in openmpi, which produces fake info. between process in mpi, so info. like rank, size are all incorrect. Hope this can help you.

**Comment by thomas-bouvier** [comment]
Experiencing the exact same issue, with the same output as @surak on one node having 8 GPUs.
